### PR TITLE
gh-111159: Fix `SyntaxError` doctests for non-builtin exception classes

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-111159

### CODE REVIEW

File: Lib/doctest.py [hunk 1]
```diff
@@ -1402,7 +1402,12 @@ def __run(self, test, compileflags, out):
                     exc_msg_index = next(
                         index
                         for index, line in enumerate(formatted_ex)
-                        if line.startswith(f"{exception[0].__name__}:")
+                        if (
+                            line.startswith(f"{exception[0].__qualname__}:")
+                            or line.startswith(
+                                f"{exception[0].__module__}.{exception[0].__qualname__}:",
+                            )
+                        )
```
Comment (scoder) [inline Lib/doctest.py]: Given that we might pass an arbitrary amount of lines before we find the final exception line (almost always the last line in the stack trace), I think it's worth preparing the two constant f-string values outside of the loop. Note that you can pass a tuple of strings into `str.startswith()`.

**Review by scoder — APPROVED** [review]
Looks almost good to me, just a performance+readability remark.

File: Lib/doctest.py [hunk 2]
```diff
@@ -1399,10 +1399,12 @@ def __run(self, test, compileflags, out):
                     # we don't care about the carets / suggestions / etc
                     # We only care about the error message and notes.
                     # They start with `SyntaxError:` (or any other class name)
+                    exc_name = f"{exception[0].__qualname__}:"
+                    exc_fullname = f"{exception[0].__module__}.{exception[0].__qualname__}:"
```
Comment (scoder) [inline Lib/doctest.py]:
```suggestion
                    exception_line_prefixes = (
                        f"{exception[0].__qualname__}:",
                        f"{exception[0].__module__}.{exception[0].__qualname__}:",
                    )
```

File: Lib/doctest.py [hunk 3]
```diff
@@ -1399,10 +1399,12 @@ def __run(self, test, compileflags, out):
...
                     exc_msg_index = next(
                         index
                         for index, line in enumerate(formatted_ex)
-                        if line.startswith(f"{exception[0].__name__}:")
+                        if line.startswith(exc_name) or line.startswith(exc_fullname)
```
Comment (scoder) [inline Lib/doctest.py]:
```suggestion
                        if line.startswith(exception_line_prefixes)
```

**Review by scoder — COMMENTED** [review]
What about this?

File: Lib/test/test_doctest.py [hunk 4]
```diff
@@ -3310,6 +3310,25 @@ def test_syntax_error_with_note(cls, multiline=False):
     raise exc
 
 
+def test_syntax_error_subclass_from_stdlib():
+    """
+    `ParseError` is a subclass of `SyntaxError`, but it is not a builtin:
+
+    >>> from xml.etree.ElementTree import ParseError
+    >>> test_syntax_error_subclass_from_stdlib()
```
Comment (sobolevn) [inline Lib/test/test_doctest.py]:
```suggestion
    >>> test_syntax_error_subclass_from_stdlib()
```

**Comment by scoder** [comment]
Looks good now.
This fixes a bug introduced in 3.13a1. Not sure what the current policy is – would that suggest adding a news entry for 3.13a2?

**Review by iritkatriel — APPROVED** [review]

**Comment by sobolevn** [comment]
Thanks everyone! 👍

**Pull request merged**